### PR TITLE
bump-formula-pr dry-run should reflect new version

### PR DESCRIPTION
**Pull request by ilovezfs**
- simulate version change for dry-run
- make sure we're using :devel version if called with --devel

### CODE REVIEW

**Comment by ilovezfs** [comment]
@MikeMcQuaid This addresses https://github.com/Homebrew/brew/issues/229

File: Library/Homebrew/formulary.rb [hunk 1]
```diff
@@ -36,8 +36,12 @@ def self.load_formula(name, path, contents, namespace)
     end
   end
 
-  def self.load_formula_from_path(name, path)
-    contents = path.open("r") { |f| set_encoding(f).read }
+  def self.formula_contents_from_path(path)
+    contents = path.open("r") { |f| set_encoding(f).read } unless contents
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: Is the `unless` a typoe; that variable is never set here?

Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: I'm pretty 👎 on modifying `formulary` to improve the output of a `--dry-run` option just used by developers.

Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: Could maybe just write it out to a temporary file or use StringIO instead?

Comment (ilovezfs) [inline Library/Homebrew/formulary.rb]: Yup, came from before I factored out formula_contents_from_path.

Comment (ilovezfs) [inline Library/Homebrew/formulary.rb]: The point was to avoid copying the set_encoding logic.

Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: Cool. Feels like StringIO may be your best fit here.

**Comment by ilovezfs** [comment]
@MikeMcQuaid PR refreshed to use a temp file.

**Comment by MikeMcQuaid** [comment]
👍 assuming StringIO was tried and doesn't work.

**Comment by ilovezfs** [comment]
@MikeMcQuaid Would prefer to use StringIO, but Utils::Inreplace expects a path, and StringIO's .path is nil. I could certainly use something other than Utils::Inreplace but then the dry-run would be less realistic. Which approach do you prefer?

**Comment by MikeMcQuaid** [comment]
@ilovezfs I think it'd be reasonable to fix `Utils::Inreplace` so it can work with `StringIO`; just want to avoid fiddling with Formula loading stuff.

**Comment by ilovezfs** [comment]
Sure, that sounds useful. I'll take a look.

**Comment by ilovezfs** [comment]
@MikeMcQuaid PR refreshed to use StringIO.

**Comment by xu-cheng** [comment]
Can you use `Formulary.from_contents` directly ?

**Comment by ilovezfs** [comment]
@xu-cheng Possibly. I'll try that.

**Comment by ilovezfs** [comment]
@xu-cheng OK, updated to use `Formulary.from_contents`

**Comment by ilovezfs** [comment]
@MikeMcQuaid @xu-cheng PR refreshed to avoid changing `inreplace` at all.

**Comment by ilovezfs** [comment]
Any changes desired?

**Comment by MikeMcQuaid** [comment]
LGTM 👍

**Comment by ilovezfs** [comment]
@UniqMartin Do you mind looking this over as well?

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 2]
```diff
@@ -51,32 +51,37 @@ def bump_formula_pr
...
+      resource_specs = formula_spec.specs
+      old_tag = resource_specs[:tag]
+      old_revision = resource_specs[:revision]
+      ohai "replace '#{old_tag}' with '#{new_tag}'" if ARGV.dry_run?
+      s.gsub!(/['"]#{old_tag}['"]/, "\"#{new_tag}\"")
+      ohai "replace '#{old_revision}' with '#{new_revision}'" if ARGV.dry_run?
+      s.gsub!(old_revision, new_revision)
+    end
+
+    raise Utils::InreplaceError.new({ formula.path => s.errors }) if s.errors.any?
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Use `raise Utils::InreplaceError.new(formula.path => s.errors)` instead. It's preferable to omit the curly braces and the surrounding spaces. (Make sure to check for other style violations I might have overlooked via `brew style <file>`.)

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Any thoughts on adding a dryrun argument to the inreplace function itself (defaulting to dryrun =false)?

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 3]
```diff
@@ -51,32 +51,37 @@ def bump_formula_pr
...
-        old_revision = resource_specs[:revision]
-        if ARGV.dry_run?
-          ohai "replace '#{old_tag}' with '#{new_tag}'"
-          ohai "replace '#{old_revision}' with '#{new_revision}'"
-        else
-          s.gsub!(/['"]#{old_tag}['"]/, "\"#{new_tag}\"")
-          s.gsub!(old_revision, new_revision)
-        end
-      end
+    s = (formula.path.open("r") { |f| Formulary.set_encoding(f).read }).extend(StringInreplaceExtension)
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Consider splitting this into two lines for readability. Repeating the `s` doesn't feel like a burden and avoiding the awkward parentheses seems like a win to me.

```
s = formula.path.open("r") { |f| Formulary.set_encoding(f).read }
s.extend(StringInreplaceExtension)
```

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: More duplicative code that would be avoided if inreplace had a dryrun argument.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 4]
```diff
@@ -51,32 +51,37 @@ def bump_formula_pr
...
+      s.gsub!(/['"]#{old_tag}['"]/, "\"#{new_tag}\"")
+      ohai "replace '#{old_revision}' with '#{new_revision}'" if ARGV.dry_run?
+      s.gsub!(old_revision, new_revision)
+    end
+
+    raise Utils::InreplaceError.new({ formula.path => s.errors }) if s.errors.any?
+
+    new_formula_version = Formulary.from_contents(formula.name, formula.path, s, requested_spec).version
+
+    if old_formula_version == new_formula_version
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: I'd probably make this check `old_formula_version >= new_formula_version` instead or handle that as two separate cases (`==` and `>`). This will avoid accidentally downgrading a formula, particularly in sometimes surprising cases like `Version.new("1.0.2a") < Version.new("1.0.2")`.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 5]
```diff
@@ -51,32 +51,37 @@ def bump_formula_pr
...
-        if ARGV.dry_run?
-          ohai "replace '#{old_tag}' with '#{new_tag}'"
-          ohai "replace '#{old_revision}' with '#{new_revision}'"
-        else
-          s.gsub!(/['"]#{old_tag}['"]/, "\"#{new_tag}\"")
-          s.gsub!(old_revision, new_revision)
-        end
-      end
+    s = (formula.path.open("r") { |f| Formulary.set_encoding(f).read }).extend(StringInreplaceExtension)
+    old_formula_version = Formulary.from_contents(formula.name, formula.path, s, requested_spec).version
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Can this line be made shorter (and thus more readable)? Maybe with a helper method (as the same construct is used once more below) or by simply wrapping the line a bit:

```ruby
old_formula_version = Formulary.from_contents(
  formula.name, formula.path, s, requested_spec
).version
```

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: We can do old_formula = ... on one line and then old_formula_version = old_formula.version on the next maybe?

**Comment by UniqMartin** [comment]
Probably not originally intended for this PR, but an automatic reset of the formula revision (if present) seems like something worth adding to a heavily automated process like this. Stripping lines that match

```ruby
/^  revision \d+$/
```

will probably work just fine in almost all cases.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 4 above)
Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: I wonder if this should be encapsulated in the Version class with a method like .validate_version_is_new that raises its own exception type?

**Comment by ilovezfs** [comment]
Yeah I already have code for that locally ;)

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 2 above)
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: It doesn't seem like a dry-run mode would be that useful for our primary use case in formulae. And given that subsequent replacements might depend on each other (including situations where order matters), the implementation of that would be slightly more than just trivial.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 4 above)
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: I don't mind having nice abstractions, but it would be nice to have more than one client (currently this command only) who will use it. I think a dedicated exception class for this kind of error might be useful (e.g. for use in `brew audit`), but hiding `raise BazException if foo < bar` behind `Version#validate_version_is_new` seems a bit too much to me and actually feels less readable.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 2 above)
Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: I'm unhappy with a couple things in this PR. 1) It's not just using the standard inreplace function in the non-dry-run case 2) the multiple inline `if ARGV.dry_run?` checks 3) fixing either of those two issues would seem to require duplication of the specific replacements in a separate dry-run block. I'm leaning toward (3) as the least bad option if inreplace itself cannot have a dryrun parameter. But perhaps you have another idea?

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 5 above)
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Works for me. Though since `old_formula_version` is exactly as long as `old_formula.version`, using `old_formula.version` directly seems more desirable than creating a separate variable.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 4 above)
Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: I don't think I'd hide. It would be begin ... rescue and then a specific message about needing to bump the formula manually.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 5 above)
Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Right, but in that case I think all uses of new_formula_version in the file should also be replaced with new_formula.version

Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Of course that's what I implied, though that might be undesirable from the perspective of keeping the differences small (and avoiding changing unrelated code).

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: which is precisely the reason I did it the way it is here :)

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: I think I prefer your original suggestion above unless we're also changing new_formula_version to new_formula.version

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 2 above)
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: My inclination would be to create a helper function that receives a list of replacement pairs and then uses that information to print the `ohai` lines and to make the actual replacements. It wouldn't even need a dry-run flag, as I think it's useful to print those messages when not in dry-run mode.

The only difference in dry-run mode would be that the modified file contents are not written to disk.

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Would you accumulate the pairs in an array of arrays and then do the actual replacement in a single call to `inreplace ... do` looping through them, or would you open the file multiple multiple times with multiple `inreplace` calls, and then revert the file with `git checkout --` if something goes wrong?

Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: I'd most certainly do the former. Modifying a file on disk and then reverting the changes with Git sounds appropriate for a human, but less so for a piece of code that should be perfectly able to do the modification in memory.

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Ah OK. I figured that's what you meant, but wasn't sure since it sounded like the helper function would be doing an actual replace upon each invocation.

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Oh OK, I see you said "a list of replacement pairs," so you're imagining the accumulation occurring outside of the helper itself. Fair enough.

**Comment by ilovezfs** [comment]
@UniqMartin PR refreshed with changes discussed.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 6]
```diff
@@ -15,6 +15,39 @@
...
+    else
+      Utils::Inreplace.inreplace(formula.path) do |s|
+        replacement_pairs.each do |old, new|
+          ohai "replace \"#{old}\" with \"#{new}\"" unless ARGV.flag?("--quiet")
+          s.gsub!(old, new)
+        end
+      end
+      formula.path.open("r") { |f| Formulary.set_encoding(f).read }
+    end
+  end
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: This duplication between the two branches of the `if` bothers me. Why not just take the first half and squeeze in the following line before returning `s` and get rid of the `else` branch?

```ruby
formula.path.atomic_write(s) unless ARGV.dry_run?
```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 7]
```diff
@@ -15,6 +15,39 @@
 require "formula"
 
 module Homebrew
+  def inreplace_pairs(formula, replacement_pairs)
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: It would feel more appropriate if the first argument were a `Pathname` instance. This method doesn't need to know that it is operating on a formula (and it could as well operate on an arbitrary file).

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 6 above)
Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Because I'd prefer to use the "real" inreplace for the non-dry-run. That was my main gripe with the way it was before this.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 8]
```diff
@@ -51,32 +84,31 @@ def bump_formula_pr
...
+      replacement_pairs << [formula_spec.url, new_url]
+      replacement_pairs << [old_hash, new_hash]
+    else
+      replacement_pairs << [formula_spec.specs[:tag], new_tag]
+      replacement_pairs << [formula_spec.specs[:revision], new_revision]
     end
 
-    new_formula = Formulary.load_formula_from_path(formula.name, formula.path)
-    new_formula_version = new_formula.version
+    contents = inreplace_pairs(formula, replacement_pairs)
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Maybe `new_contents` or even `new_formula_code` for clarity?

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: new_contents sounds good.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 9]
```diff
@@ -51,32 +84,31 @@ def bump_formula_pr
...
+    old_formula_version = formula_version(formula, requested_spec)
+
+    replacement_pairs = []
+    if new_url_hash
+      replacement_pairs << [formula_spec.url, new_url]
+      replacement_pairs << [old_hash, new_hash]
+    else
+      replacement_pairs << [formula_spec.specs[:tag], new_tag]
+      replacement_pairs << [formula_spec.specs[:revision], new_revision]
     end
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: This could be simplified to (it gets a bit longer, but less repetition):

```ruby
replacement_pairs = if new_url_hash
  [
    [formula_spec.url, new_url],
    [old_hash, new_hash],
  ]
else
  [
    [formula_spec.specs[:tag], new_tag],
    [formula_spec.specs[:revision], new_revision],
  ]
end
```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 6 above)
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: What do we gain from using the “real” `inreplace` that offsets this duplication?

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: The confidence in knowing that the non-dry-run path will automatically get whatever changes are made to inreplace in the future. But this duplication is exactly why I was saying I wish inreplace itself had a dry-run mode since this whole function feels like a big hack around that deficiency,

Comment (ilovezfs) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: @UniqMartin I would feel much more comfortable with removing the duplication if we moved this function into Utils::Inreplace, so that it would be "guaranteed" to be updated appropriately whenever Utils::Inreplace changes.

**Comment by ilovezfs** [comment]
@UniqMartin perhaps we should return to @MikeMcQuaid's idea to support `StringIO` class in `Utils::Inreplace`. Here's what that looked like when it was in the PR:
```
commit 8de273e6067937731374ee41edf69ebf07cd039e
Author: ilovezfs <ilovezfs@icloud.com>
Date:   Thu Jun 9 02:54:08 2016 -0700

    Utils::Inreplace: support StringIO
    
    This allows `inreplace` to accept paths or StringIOs or both.

diff --git a/Library/Homebrew/utils/inreplace.rb b/Library/Homebrew/utils/inreplace.rb
index ee47da4..20bfd1b 100644
--- a/Library/Homebrew/utils/inreplace.rb
+++ b/Library/Homebrew/utils/inreplace.rb
@@ -15,26 +15,43 @@ module Utils
     # HOMEBREW_PREFIX is available in the embedded patch.
     # inreplace supports regular expressions.
     # <pre>inreplace "somefile.cfg", /look[for]what?/, "replace by #{bin}/tool"</pre>
-    def inreplace(paths, before = nil, after = nil)
-      errors = {}
 
-      Array(paths).each do |path|
-        s = File.open(path, "rb", &:read).extend(StringInreplaceExtension)
+    def inreplace_impl(input, before, after, errors, &block)
+      file = input.class <= Pathname || input.class <= String
+      s = file ?  File.open(input, "rb", &:read) : input.read
+      s = s.extend(StringInreplaceExtension)
 
-        if before.nil? && after.nil?
-          yield s
-        else
-          after = after.to_s if Symbol === after
-          s.gsub!(before, after)
-        end
+      if before.nil? && after.nil?
+        yield s
+      else
+        after = after.to_s if Symbol === after
+        s.gsub!(before, after)
+      end
+
+      errors[input] = s.errors if s.errors.any?
 
-        errors[path] = s.errors if s.errors.any?
+      if file
+        Pathname.new(input).atomic_write(s)
+      else
+        input.rewind
+        input.write(s)
+        input.rewind
+      end
+    end
 
-        Pathname(path).atomic_write(s)
+    def inreplace(input, before = nil, after = nil, &block)
+      errors = {}
+      if input.class <= Array
+        input.each do |i|
+          inreplace_impl(i, before, after, errors, &block)
+        end
+      else
+        inreplace_impl(input, before, after, errors, &block)
       end
 
       raise InreplaceError.new(errors) if errors.any?
     end
     module_function :inreplace
+    module_function :inreplace_impl
   end
 end

```

**Comment by UniqMartin** [comment]
@ilovezfs Before I come forward with a new suggestion or more complaints about code duplication, let me summarize what I learned so far. Ideally, you'd like to have an implementation of `inreplace` that can everything it currently does and in addition:

- Returns the resulting string to the caller after all substitutions have been applied.
- Supports a dry-run/read-only mode where the resulting string is not written back to the input file.

Is that correct? I believe `StringIO` will allow for even more flexibility, but is any of its flexibility needed or was it just a tool to achieve the above goals? Will `StringIO` enable additional usage patterns and do you see use cases for them? (Just trying to finally understand the problem and the constraints better.)

**Comment by ilovezfs** [comment]
@UniqMartin yes that all sounds accurate. Only tweak I'd make to what you said is that I wouldn't necessarily care about getting the string itself back if we could operate in a dry-run fashion on the contents of a formula without having to re-feed the Formulary a new copy of the contents, but otherwise, yes.

I don't like the idea of inventing synthetic use cases whose necessity has not yet presented itself just to justify a new feature, but of course boneyard-formula-pr would use it.

**Comment by MikeMcQuaid** [comment]
> Returns the resulting string to the caller after all substitutions have been applied.
> Supports a dry-run/read-only mode where the resulting string is not written back to the input file.

At this point is it not worth just using `String#gsub`?

**Comment by ilovezfs** [comment]
@MikeMcQuaid It is: https://github.com/Homebrew/brew/pull/318/files#diff-bd1e2ccd6350a553602c33a962d60309R24

**Comment by MikeMcQuaid** [comment]
@ilovezfs Sorry, I was commenting on @UniqMartin's `inreplace` comments.

**Comment by ilovezfs** [comment]
@MikeMcQuaid In any case, I'm happy with this as it currently is written in the because the non-dry-run case uses standard `inreplace`. Yes, there are ten lines of "duplication" as @UniqMartin has pointed out, but it seems that's outweighed by the fact that the non-dry-run case gets to use the standard function. If, however, it is preferred to just do `formula.path.atomic_write(s) unless ARGV.dry_run?` and thereby avoid the duplication at the cost of not using the "standard" function in the non-dry-run case, I'm happy to change to that too, as it's what was already in the PR two iterations ago.

Ideally, inreplace would natively support this sort of thing, but if that's not immediately obvious to all involved, I'll live with this functionality sitting off in a random helper function in an obscure dev-cmd as long as it gets the job done. :)

**Comment by MikeMcQuaid** [comment]
:shipit:

**Comment by ilovezfs** [comment]
@MikeMcQuaid Done.

**Comment by UniqMartin** [comment]
Thanks for taking the time to explain the use case and your thoughts behind using `inreplace` even at the cost of a few more lines. I think extending `inreplace` with a dry-run mode makes sense, even if the majority of use cases (formulae) won't need that. But that can happen in a separate PR and then applied to this command. In case the message got lost: Great work in this PR and definitely an improvement! :tada:
